### DISCUSSION TITLE
How to block bg lifespan task until app can serve?

### DISCUSSION THREAD

**Discussion by emanueljg** [fastapi/fastapi#11453] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def lifespan():
  asyncio.create_task(bg_task())
  yield

async def bg_task():
  await app_is_started()
  resp = requests.get("127.0.0.1:8000")
```

### Description

I run a task in my lifespan startup in the background. I want my lifespan to then finish and be able to serve requests, and for my task to after this become awakened and run to completion.  If I have no such blocking, the task executes *during* the lifespan execution by which point my application is not ready to serve this task, meaning I get a 404 and the lifespan crashes. I need some way of blocking this task until my app is ready.

Of course, I could do something akin to a repetitive while-loop until the request goes through, but that feels very ugly and I'd like something more idiomatic and elegant.

### Operating System

Linux

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

Python 3.11.8

**Comment by YuriiMotov** [comment, marked as answer]
Hi!

Are you going to run only one worker?
If you run several workers you will have to wait until all of them are started. Otherwise you can still get 404 error.

Also, it will not work with this implementation, since your `requests.get("127.0.0.1:8000")` call is not async and it will block event loop and your app will be frozen.
You need either to use async library to perform requests (e.g. [httpx](https://www.python-httpx.org/async/)) or run this task in the thread (not as async task)

**Comment by emanueljg** [comment]
> Are you going to run only one worker? If you run several workers you will have to wait until all of them are started. Otherwise you can still get 404 error.

Currently only one worker but planning on extending. How do I do either or?

> Also, it will not work with this implementation, since your `requests.get("127.0.0.1:8000")` call is not async and it will block event loop and your app will be frozen. You need either to use async library to perform requests (e.g. [httpx](https://www.python-httpx.org/async/)) or run this task in the thread (not as async task)

Yeah I know, I use aiohttp in my app. I just used `requests` to get the point across of an HTTP request being made to my own API.

**Reply by YuriiMotov** [reply]
I would implement this logic as a while-loop. Something like:
```
import asyncio
from contextlib import asynccontextmanager
import httpx
from fastapi import FastAPI

async def bg_task():
    async with httpx.AsyncClient() as client:
        while True:
            try:   
                r = await client.get('http://127.0.0.1:8000/')
                if r.status_code == 200:
                    break
            except httpx.RequestError as e:
                pass
        print(f"{r.text=}")

@asynccontextmanager
async def lifespan(app: FastAPI):
    task = asyncio.create_task(bg_task())
    yield

app = FastAPI(lifespan=lifespan)

@app.get('/')
async def r1():
    return "123"
```

In this implementation this task will be executed on each worker. To run it only on one worker you need to use any mechanism like Lock or Semaphore.
